### PR TITLE
c#, fix: class names + namespace conflicts

### PR DESCRIPTION
**Pull request by dcb6**
C# generator changes:
* Fixes a bug where conflicting class names and namespaces cause compile to fail.

Seed changes:
* Prefixing a test with the language causes it to only be run for that language. Logic [here](https://github.com/fern-api/fern/pull/4229/files#diff-7bdcadc7fd64039486bff458d0080bf16b32332d5de974e3584859da56f014d3R11)

### CODE REVIEW

File: seed/csharp-sdk/csharp-namespace-conflict/src/SeedCsharpNamespaceConflict/A/Types/B.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace SeedCsharpNamespaceConflict.A;
+
+public record B
+{
+    /// <summary>
+    /// This reference to A should not use it's fully qualified name because it's in the same package as A.
+    /// </summary>
+    [JsonPropertyName("id")]
+    public required A Id { get; set; }
```
Comment (dcb6) [inline seed/csharp-sdk/csharp-namespace-conflict/src/SeedCsharpNamespaceConflict/A/Types/B.cs:13]: we're in the same namespace as `A` so we don't need to fully qualify

File: seed/csharp-sdk/csharp-namespace-conflict/src/SeedCsharpNamespaceConflict.Test/Unit/Serialization/ATest.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  ""name"": ""foo""
+}
+";
+
+        var serializerOptions = new JsonSerializerOptions
+        {
+            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
+        };
+
+        var deserializedObject = JsonSerializer.Deserialize<SeedCsharpNamespaceConflict.A.A>(
```
Comment (dcb6) [inline seed/csharp-sdk/csharp-namespace-conflict/src/SeedCsharpNamespaceConflict.Test/Unit/Serialization/ATest.cs:28]: Fully qualified example

File: packages/seed/src/commands/test/testWorkspaceFixtures.ts [hunk 3]
```diff
@@ -7,7 +7,8 @@ import { GeneratorWorkspace } from "../../loadGeneratorWorkspaces";
 import { printTestCases } from "./printTestCases";
 import { TestRunner } from "./test-runner";
 
-export const FIXTURES_TO_IGNORE = ["sever-sent-event-examples", "server-sent-event-examples"];
+export const FIXTURES_TO_IGNORE = ["server-sent-event-examples", "server-sent-events"];
+export const LANGUAGE_SPECIFIC_FIXTURE_PREFIXES = ["csharp", "go", "java", "python", "ruby", "ts"];
```
Comment (dsinghvi) [inline packages/seed/src/commands/test/testWorkspaceFixtures.ts:11]: nice!

File: generators/csharp/codegen/src/context/AbstractCsharpGeneratorContext.ts [hunk 4]
```diff
@@ -30,6 +31,7 @@ export abstract class AbstractCsharpGeneratorContext<
     public readonly project: CsharpProject;
     public readonly csharpTypeMapper: CsharpTypeMapper;
     public publishConfig: FernGeneratorExec.NugetGithubPublishInfo | undefined;
+    private baseNamespacesCache: Set<String> | null = null;
```
Comment (dsinghvi) [inline generators/csharp/codegen/src/context/AbstractCsharpGeneratorContext.ts:34]: nit: Set<string> | undefined; 

Rest of the code base tends to prefer `undefined` over `null`

File: generators/csharp/codegen/src/project/CSharpFile.ts [hunk 5]
```diff
@@ -10,12 +11,17 @@ export declare namespace CSharpFile {
...
         /* Directory of the filepath */
         directory: RelativeFilePath;
+        context: AbstractCsharpGeneratorContext<any>;
     }
 }
 
 export class CSharpFile extends File {
-    constructor({ clazz, directory }: CSharpFile.Args) {
-        super(`${clazz.name}.cs`, directory, clazz.toString());
+    constructor({ clazz, directory, context }: CSharpFile.Args) {
```
Comment (dsinghvi) [inline generators/csharp/codegen/src/project/CSharpFile.ts:19]: by requiring `context` this makes it pretty hard to unit test `CsharpFile`. Can we insted take in the values separately. So arguments for `namespace`, `baseNamespaces`, etc.

Comment (dcb6) [reply generators/csharp/codegen/src/project/CSharpFile.ts:19]: yeah will do

File: generators/csharp/codegen/src/ast/ClassReference.ts [hunk 6]
```diff
@@ -26,8 +26,16 @@ export class ClassReference extends AstNode {
     }
 
     public write(writer: Writer): void {
-        writer.addReference(this);
-        writer.write(`${this.name}`);
+        const qualifiedTypeNameRequired =
+            writer.getAllBaseNamespaces().has(this.name) &&
+            // qualified type name not required if we're in the same base namespace as the conflicting type
+            !writer.getNamespace()?.startsWith(`${writer.getRootNamespace()}.${this.name}`);
```
Comment (dsinghvi) [inline generators/csharp/codegen/src/ast/ClassReference.ts:32]: can we move this into a helper function and specifically document the edge cases

**Review by dsinghvi — APPROVED** [review]

**Pull request merged**